lib/net/pool.js

```diff
@@ -2012,6 +2123,99 @@ class Pool extends EventEmitter {
     peer.sendCFCheckpt(packet.filterType, packet.stopHash, filterHeaders);
   }
 
+  /**
+   * Handle peer `CFCheckpt` packet.
+   * @method
+   * @private
+   * @param {Peer} peer
+   * @param {CFCheckptPacket} packet
+   */
+
+  async handleCFCheckpt(peer, packet) {
+    if (!this.options.neutrino) {
+      peer.ban();
+      peer.destroy();
+    }
+  }
+
+  /**
+   * Handle peer `CFHeaders` packet.
+   * @method
+   * @private
+   * @param {Peer} peer - Sender.
+   * @param {CFHeadersPacket} packet - Packet to handle.
+   * @returns {void}
+   */
+  async handleCFHeaders(peer, packet) {
+    this.logger.info('Received CFHeaders packet from %s', peer.hostname());
+    if (!this.options.neutrino) {
+      peer.ban();
+      peer.destroy();
+      return;
+    }
+
+    const filterType = packet.filterType;
+    assert(filterType === this.getcfheadersFilterType);
+    const stopHash = packet.stopHash;
+    assert(stopHash.equals(this.getcfheadersStopHash));
+    let previousFilterHeader = packet.previousFilterHeader;
+    const filterHashes = packet.filterHashes;
+    assert(filterHashes.length <= 2000);
+    let blockHeight = await this.chain.getHeight(stopHash)
+      - filterHashes.length;
+    const stopHeight = await this.chain.getHeight(stopHash);
+    for (const filterHash of filterHashes) {
+      assert(blockHeight < stopHeight);
+      const basicFilter = new BasicFilter();
+      basicFilter._hash = filterHash;
+      const filterHeader = basicFilter.header(previousFilterHeader);
+      const lastFilterHeader = this.cfHeaderChain.tail;
+      const cfHeaderEntry = new CFHeaderEntry(
+        filterHash, lastFilterHeader.height + 1);
+      this.cfHeaderChain.push(cfHeaderEntry);
+      // todo: verify the filterHeader
+      // todo: save the filterHeader
+      previousFilterHeader = filterHeader;
+      this.chain.db.neutrinoState.headerHeight = blockHeight;
```

Comment:
I don't think pool should ever call anything in `chain.db`. There should probably helper functions in `this.chain` instead that manage the incoming data and just tell pool if its valid (do we need to ban the peer that sent it to us, etc). See `pool.addBlock()` for example